### PR TITLE
Lbgpu node vel

### PR DESCRIPTION
**Pull request by fweik**
Description of changes:
 - Factored out velocity getter, and obey boundary velocity for
   both interpolation schemes,
- Removed some globals
- Leak less

### CODE REVIEW

**Comment by fweik** [comment]
@mkuron do you have seen the rocm error before? (`..### HCC STATUS_CHECK Error: HSA_STATUS_ERROR_INVALID_ISA (0x100f) at file:mcwamp_hsa.cpp line:1193`, looks like a toolchain issue to me)

**Comment by mkuron** [comment]
No idea. Could be some variation of an out-of-memory error or out-of-registers.

**Comment by KaiSzuttor** [comment]
```
/**
   * The instruction set architecture is invalid.
   */
  HSA_STATUS_ERROR_INVALID_ISA = 0x100F,
```

https://github.com/RadeonOpenCompute/ROCR-Runtime/blob/faa3ae51388517353afcdaf9c16621f879ef0a59/src/inc/hsa.h#L196-L199

**Comment by KaiSzuttor** [comment]
actually I don't understand why the compilation is not terminated, there are a number of compile errors:
``` bash
/builds/espressomd/espresso/src/core/electrostatics_magnetostatics/p3m_gpu_error_cuda.cu:104:37: error:  'Utils::sinc':  no overloaded function has restriction specifiers that are compatible with the ambient context 'p3m_k_space_error_gpu_kernel_ik'
        int_pow<2 * cao>(Utils::sinc(meshi.x * nx) * Utils::sinc(meshi.y * ny) *
                                    ^
/builds/espressomd/espresso/src/core/electrostatics_magnetostatics/p3m_gpu_error_cuda.cu:104:37: error:  'Utils::sinc':  no overloaded function has restriction specifiers that are compatible with the ambient context 'p3m_k_space_error_gpu_kernel_ik'
/builds/espressomd/espresso/src/core/electrostatics_magnetostatics/p3m_gpu_error_cuda.cu:104:65: error:  'Utils::sinc':  no overloaded function has restriction specifiers that are compatible with the ambient context 'p3m_k_space_error_gpu_kernel_ik'
        int_pow<2 * cao>(Utils::sinc(meshi.x * nx) * Utils::sinc(meshi.y * ny) *
                                                                ^
/builds/espressomd/espresso/src/core/electrostatics_magnetostatics/p3m_gpu_error_cuda.cu:104:65: error:  'Utils::sinc':  no overloaded function has restriction specifiers that are compatible with the ambient context 'p3m_k_space_error_gpu_kernel_ik'
/builds/espressomd/espresso/src/core/electrostatics_magnetostatics/p3m_gpu_error_cuda.cu:105:37: error:  'Utils::sinc':  no overloaded function has restriction specifiers that are compatible with the ambient context 'p3m_k_space_error_gpu_kernel_ik'
                         Utils::sinc(meshi.z * nz));
                                    ^
/builds/espressomd/espresso/src/core/electrostatics_magnetostatics/p3m_gpu_error_cuda.cu:105:37: error:  'Utils::sinc':  no overloaded function has restriction specifiers that are compatible with the ambient context 'p3m_k_space_error_gpu_kernel_ik'
/builds/espressomd/espresso/src/core/electrostatics_magnetostatics/p3m_gpu_error_cuda.cu:151:39: error:  'Utils::sinc':  no overloaded function has restriction specifiers that are compatible with the ambient context 'p3m_k_space_error_gpu_kernel_ad'
          U2 = pow((double)Utils::sinc(meshi.x * nmx) *
                                      ^
/builds/espressomd/espresso/src/core/electrostatics_magnetostatics/p3m_gpu_error_cuda.cu:151:39: error:  'Utils::sinc':  no overloaded function has restriction specifiers that are compatible with the ambient context 'p3m_k_space_error_gpu_kernel_ad'
/builds/espressomd/espresso/src/core/electrostatics_magnetostatics/p3m_gpu_error_cuda.cu:152:35: error:  'Utils::sinc':  no overloaded function has restriction specifiers that are compatible with the ambient context 'p3m_k_space_error_gpu_kernel_ad'
                       Utils::sinc(meshi.y * nmy) * Utils::sinc(meshi.z * nmz),
                                  ^
/builds/espressomd/espresso/src/core/electrostatics_magnetostatics/p3m_gpu_error_cuda.cu:152:35: error:  'Utils::sinc':  no overloaded function has restriction specifiers that are compatible with the ambient context 'p3m_k_space_error_gpu_kernel_ad'
/builds/espressomd/espresso/src/core/electrostatics_magnetostatics/p3m_gpu_error_cuda.cu:152:64: error:  'Utils::sinc':  no overloaded function has restriction specifiers that are compatible with the ambient context 'p3m_k_space_error_gpu_kernel_ad'
                       Utils::sinc(meshi.y * nmy) * Utils::sinc(meshi.z * nmz),
                                                               ^
/builds/espressomd/espresso/src/core/electrostatics_magnetostatics/p3m_gpu_error_cuda.cu:152:64: error:  'Utils::sinc':  no overloaded function has restriction specifiers that are compatible with the ambient context 'p3m_k_space_error_gpu_kernel_ad'
/builds/espressomd/espresso/src/core/electrostatics_magnetostatics/p3m_gpu_error_cuda.cu:210:39: error:  'Utils::sinc':  no overloaded function has restriction specifiers that are compatible with the ambient context 'p3m_k_space_error_gpu_kernel_ik_i'
          U2 = pow((double)Utils::sinc(meshi.x * nmx) *
                                      ^
/builds/espressomd/espresso/src/core/electrostatics_magnetostatics/p3m_gpu_error_cuda.cu:210:39: error:  'Utils::sinc':  no overloaded function has restriction specifiers that are compatible with the ambient context 'p3m_k_space_error_gpu_kernel_ik_i'
/builds/espressomd/espresso/src/core/electrostatics_magnetostatics/p3m_gpu_error_cuda.cu:211:35: error:  'Utils::sinc':  no overloaded function has restriction specifiers that are compatible with the ambient context 'p3m_k_space_error_gpu_kernel_ik_i'
                       Utils::sinc(meshi.y * nmy) * Utils::sinc(meshi.z * nmz),
                                  ^
/builds/espressomd/espresso/src/core/electrostatics_magnetostatics/p3m_gpu_error_cuda.cu:211:35: error:  'Utils::sinc':  no overloaded function has restriction specifiers that are compatible with the ambient context 'p3m_k_space_error_gpu_kernel_ik_i'
/builds/espressomd/espresso/src/core/electrostatics_magnetostatics/p3m_gpu_error_cuda.cu:211:64: error:  'Utils::sinc':  no overloaded function has restriction specifiers that are compatible with the ambient context 'p3m_k_space_error_gpu_kernel_ik_i'
                       Utils::sinc(meshi.y * nmy) * Utils::sinc(meshi.z * nmz),
                                                               ^
/builds/espressomd/espresso/src/core/electrostatics_magnetostatics/p3m_gpu_error_cuda.cu:211:64: error:  'Utils::sinc':  no overloaded function has restriction specifiers that are compatible with the ambient context 'p3m_k_space_error_gpu_kernel_ik_i'
/builds/espressomd/espresso/src/core/electrostatics_magnetostatics/p3m_gpu_error_cuda.cu:273:39: error:  'Utils::sinc':  no overloaded function has restriction specifiers that are compatible with the ambient context 'p3m_k_space_error_gpu_kernel_ad_i'
          U2 = pow((double)Utils::sinc(meshi.x * nmx) *
                                      ^
fatal error: too many errors emitted, stopping now [-ferror-limit=]
20 errors generated.
```

**Comment by mkuron** [comment]
Good find. So this is actually something like an unresolved symbol error.

> actually I don't understand why the compilation is not terminated

That‘s a known hipcc bug. It doesn‘t produce nonzero exit codes if compilation failed. Usually that‘s not a problem because the object file would subsequently be missing and CMake terminates anyway, but in the current case, an (incomplete?) object file seems to have already been written. That issue was fixed last month: https://github.com/ROCm-Developer-Tools/HIP/pull/1117.

**Comment by jngrad** [comment]
Concerning the `no overloaded function has restriction specifiers` compiler error, the explanation can be found in https://github.com/ROCm-Developer-Tools/HIP/issues/374#issuecomment-384256682. The error message disappears when using `#if defined(__CUDACC__) or defined(__HIPCC__)` instead of
https://github.com/espressomd/espresso/blob/7eebdb9fa3155225fa04d92dafb7a50f55ce0c78/src/utils/include/utils/device_qualifier.hpp#L4

**Comment by fweik** [comment]
@KaiSzuttor there is a merge regression now that reintroduces the `ek_params` stuff?

**Comment by KaiSzuttor** [comment]
I'll try again...

File: src/core/grid_based_algorithms/lbgpu_cuda.cu [hunk 1]
```diff
@@ -2181,9 +2150,11 @@ __global__ void reset_boundaries(LB_boundaries_gpu boundaries) {
  *  @param[out]    n_b     Local node residing in array b
  *  @param[in,out] d_v     Local device values
  *  @param[in,out] node_f  Local node force density
+ *  @param[in]     ek_parameters_gpu  Parameters for the electrokinetics
  */
 __global__ void integrate(LB_nodes_gpu n_a, LB_nodes_gpu n_b, LB_rho_v_gpu *d_v,
                           LB_node_force_density_gpu node_f,
+                          EK_parameters *ek_parameters_gpu,
```
Comment (fweik) [inline src/core/grid_based_algorithms/lbgpu_cuda.cu]: Here e.g.

Comment (KaiSzuttor) [reply src/core/grid_based_algorithms/lbgpu_cuda.cu]: this was intended by this PR, right?

Comment (fweik) [reply src/core/grid_based_algorithms/lbgpu_cuda.cu]: No, those are not needed and were removed in the master after this branch was forked of

**Comment by KaiSzuttor** [comment]
was there before merge: 
https://github.com/espressomd/espresso/blob/e0167799c15df578b212d80d5f737609d65ea7df/src/core/grid_based_algorithms/lbgpu_cuda.cu#L2157

**Comment by fweik** [comment]
Still this reverts #2877 which it should not.

**Comment by KaiSzuttor** [comment]
puh, then maybe you should merge...

**Comment by fweik** [comment]
@KaiSzuttor

**Comment by fweik** [comment]
This is too difficult to debug for me, giving up.

File: src/core/grid_based_algorithms/lbgpu_cuda.cu [hunk 2]
```diff
@@ -2538,6 +2516,11 @@ void lb_init_boundaries_GPU(int host_n_lb_boundaries, int number_of_boundnodes,
   if (this_node != 0)
     return;
 
+  float *boundary_velocity = nullptr;
+  int *boundary_node_list = nullptr;
+  int *boundary_index_list = nullptr;
+  size_t size_of_boundindex = 0;
```
Comment (mkuron) [inline src/core/grid_based_algorithms/lbgpu_cuda.cu]: These variables shadow globals, and these globals are not even used anymore. It's possible that that was the cause of the `HSA_STATUS_ERROR_INVALID_ISA` you were seeing.

**Comment by KaiSzuttor** [comment]
@fweik it's only the rocm test that fails... maybe we should not throw away this PR just because of that

**Comment by fweik** [comment]
Well you know where to find the code. I'm done with this.

File: src/core/grid_based_algorithms/lbgpu_cuda.cu (hunk 2 above)
Comment (jngrad) [reply src/core/grid_based_algorithms/lbgpu_cuda.cu]: [`HSA_STATUS_ERROR_INVALID_ISA`](https://github.com/RadeonOpenCompute/ROCR-Runtime/blob/898beb5adbe227923b84946e9d5f6570e727d34f/src/inc/hsa.h#L192-L195): The instruction set architecture is invalid.

**Comment by mkuron** [comment]
This is where the error comes from: https://github.com/RadeonOpenCompute/ROCR-Runtime/blob/e2d1950bd8fc83dd5b8352a41b829a7c57fc1073/src/core/runtime/amd_aql_queue.cpp#L820. It supposedly means "Out of VGPRs".

On CUDA, an out-of-registers condition would cause an overflow into global memory -- slowing things down, but still working. I guess we don't have infrastructure in place that would detect this kind of performance regression on CUDA though. So it's plausible that a function requiring a very large number of registers would be broken on AMD. Adding `-Xptxas=-v`to the CUDA flags reveals a few functions that are close to the limit (63 on sm_30 and 255 on later ones) and have probably already overflowed into global memory.

I think AMD can deal with 255 VGPRs too, but it's unclear to me how to measure how many of them a kernel is using. Maybe this pull request just crosses the limit. To reduce register consumption, large kernels would need to be broken up into multiple smaller ones.

**Comment by mkuron** [comment]
It's the call to `thrust::transform(..., lb_get_interpolated_velocity_gpu<27>(...))`. `thrust::transform(..., lb_get_interpolated_velocity_gpu<8>(...))` works fine. #2982 fixes the issue.

File: src/core/grid_based_algorithms/lbgpu_cuda.cu (hunk 2 above)
Comment (mkuron) [reply src/core/grid_based_algorithms/lbgpu_cuda.cu]: That‘s just a generic error code. It is used in many places for which no specific error code has been defined.

**Comment by jngrad** [comment]
@KaiSzuttor the webhook didn't trigger CI for 81b9c33

**Review by KaiSzuttor — APPROVED** [review]

**Review by mkuron — APPROVED** [review]

**Pull request merged**